Initial config implementation

PereBal opened the pull request.

PereBal committed:
Updated gitignore

thule0 commented on bigrig/config.py:
```diff
@@ -0,0 +1,202 @@
+                f"Unable to load credentials from `{path}`, it appears to be a non-yaml file"
+            ) from None
+        except OSError as exc:
+            raise OSError(f"Unable to load credentials from `{path}`") from exc
```
Nitpick: is there a lot of point of catching and reraising the same error with lightly different message?

PereBal replied:
As discussed via slack, I tend to catch these kind of exceptions to provide more user-friendly messages, while I agree that this one in particular is just the same thing, I'd like to leave it as is for consistency with the rest of the module.

thule0 commented on bigrig/config.py:
```diff
@@ -0,0 +1,202 @@
+
+
+@dataclass
+class OriginCredentials:
```
I see only origin seems to have credentials, and naming reflects that.
All the repos should have it actually. In fact, the way it works right now is  you need password only to upload, so origin is the only one that does not require credentials :)

PereBal replied:
This is what we discussed initially, but I think it just makes sense to allow having credentials in every origin,source,target entry, I'll review and update the design doc accordingly :)

thule0 commented on bigrig/config.py:
```diff
@@ -0,0 +1,202 @@
+class ConfigEntry:
+    origin: Origin
+    source: Source
+    target: t.Dict[str, Target]
```
targetS?

roman-alexeev commented:
Can we have some tests to see what is valid?  JsonSchema is a bit of a mouthful tbh.

PereBal commented:
> Can we have some tests to see what is valid? JsonSchema is a bit of a mouthful tbh.

@roman-alexeev What would you like to cover?

thule0 commented on bigrig/config.py:
```diff
@@ -0,0 +1,229 @@
+        try:
+            with open(config_path, "rt") as fid:
+                entries = [
+                    ConfigEntry.from_dict(blob=blob) for blob in yaml.safe_load_all(fid)
```
We can have more then one?

PereBal replied:
As you proposed, yes. And I would like to point out that allowing for more than one config object at a time, lets us have multiple origins to multiple targets, etc in a fairly trivial way so I'd leave it as is.

thule0 replied:
Did I? :)
So you mean going through configs as a list and run for each of them?

PereBal replied:
yea, this is related to my questions regarding packages per origin and what not :)

PereBal replied:
removed as agreed on slack, multiple origin/target will be handled by multiple instances of the app ( for now )

PereBal committed:
Add initial config implementation.

There are 3 parts of the configuration, initialization and
usage.

Validation is handled by jsonschema, small bits and pieces
are added into the corresponding dataclasses but nothing too important.

To handle initialization a simple pattern of what logging or django do has
been choes. Force user to call a certain 'initialization' funciton at
least once in the lifetime of the interpreter before making the config itself
usable.

By doing this we ensure that chached config is not going to change
during the whole execution and that only one instance (per interpreter)
of it will exist in memory.

While this is adds a little bit more on the user, it greatly simplifies
testing.

As for usage, dataclasses and OO

PereBal committed:
Exclude packaging from mypy checks

PereBal committed:
Replace settings __getattr__ with __getattribute__ to fully controll access to attributes

PereBal committed:
Replace config entries list with a single entry
